force-app/main/shared/classes/EventHandler.cls

```diff
@@ -0,0 +1,148 @@
+        es.SH_Type__c=e.Type;
+        es.SH_Unit__c=e.CRM_Unit__c;
+        es.SH_WhatId__c=e.WhatId;
+        es.SH_WhoId__c=e.WhoId;
+        es.SH_EventId__c=e.Id;
+
+        Insert es;
+        }
+    }
+
+    private void updateEventShadow( List<Event> events ) {
+
+        Set<String> allIds = new Set<String>();
+        for (Event e : events){
+            allIds.add(e.Id);
+        }
+
+        List<EventShadow__c> listEs = [SELECT Id, SH_AccountOrgNumber__c,SH_AccountParentId__c,SH_AccountParentOrgNumber__c,SH_ActivityDateTime__c,SH_ActivityType__c,SH_CreatedById__c,SH_DateActivity__c,SH_Date__c,SH_DurationTime__c,SH_EndDateTime__c,SH_EventId__c,SH_EventSubtype__c,SH_LastModifiedById__c,SH_NoPersonalInformation__c,SH_OwnerId__c,SH_Region__c,SH_RelatedToAccount__c,SH_Service__c,SH_StartDateTime__c,SH_Status__c,SH_Subject__c,SH_TaskEvent__c,SH_Type__c,SH_Unit__c,SH_WhatId__c,SH_WhoId__c FROM EventShadow__c WHERE SH_EventId__c IN :allIds];
+
+        Map<String, EventShadow__c> eventIdToEventShadow = new Map<String, EventShadow__c>();
+        for(EventShadow__c es : listEs) {
+            eventIdToEventShadow.put(es.SH_EventId__c, es);
+        }
+        for(Event e : events) {
+            EventShadow__c esRecord = eventIdToEventShadow.get(e.Id);
+            if (esRecord != null) {
+                esRecord.SH_AccountOrgNumber__c=e.CRM_AccountOrgNumber__c;
+                esRecord.SH_AccountParentId__c=e.TAG_AccountParentId__c;
+                esRecord.SH_AccountParentOrgNumber__c=e.TAG_AccountParentOrgNumber__c;
+                esRecord.SH_ActivityDateTime__c=e.ActivityDateTime;
+                esRecord.SH_ActivityType__c=e.TAG_ActivityType__c;
+                esRecord.SH_CreatedById__c=e.CreatedById;
+                esRecord.SH_Date__c=e.ActivityDate;
+                esRecord.SH_DateActivity__c=e.Date__c;
+                esRecord.SH_DurationTime__c=e.DurationInMinutes;
+                esRecord.SH_EndDateTime__c=e.EndDateTime;
+                esRecord.SH_EventSubtype__c=e.EventSubtype;
+                esRecord.SH_LastModifiedById__c=e.LastModifiedById;
+                esRecord.SH_NoPersonalInformation__c=e.TAG_NoPersonInformation__c;
+                esRecord.SH_OwnerId__c=e.OwnerId;
+                esRecord.SH_Region__c=e.CRM_Region__c;
+                esRecord.SH_RelatedToAccount__c=e.RelatedToAccount__c;
+                esRecord.SH_Service__c=e.TAG_service__c;
+                esRecord.SH_StartDateTime__c=e.StartDateTime;
+                esRecord.SH_Status__c=e.TAG_Status__c;
+                esRecord.SH_Subject__c=e.Subject;
+                esRecord.SH_TaskEvent__c=e.TaskEvent__c;
+                esRecord.SH_Type__c=e.Type;
+                esRecord.SH_Unit__c=e.CRM_Unit__c;
+                esRecord.SH_WhatId__c=e.WhatId;
+                esRecord.SH_WhoId__c=e.WhoId;
+                esRecord.SH_EventId__c=e.Id;
+            }
+            Update esRecord;
+        }
+    }
```

Comment:
```suggestion
 	private void upsertEventShadow( List<Event> events ) {

		List<EventShadow__c> eventShadows = new List<EventShadow__c>();

		for ( Event e : events ) {
			EventShadow__c es = createEventShadowObject( e );
			eventShadows.add( es );
		}

		upsert eventShadows SH_EventId__c;
	}
```

Kortes ned, og bruker til både undelete og insert, samt update.

**NB!** Krever at SH_EventId__c blir satt til følgende verdier i crm-platform-base (se bilde). Ved å benytte "Set this field as the unique record identifier from an external system", får man også lov til å skrive koden `upsert eventShadows SH_EventId__c;`, som betyr "insert eller update disse eventShadows, men bruk feltet SH_EventId__c til å matche eksisterende records. Hvis man ikke finner noen records med det feltet, lages det bare en ny automatisk.

<img width="551" alt="image" src="https://user-images.githubusercontent.com/6638390/106139896-537e1200-616e-11eb-867f-7d2c4ad9d769.png">